File: backend/app/services/pattern_recognition.py
import pandas as pd
import numpy as np
from dataclasses import dataclass


@dataclass
class PatternResult:
    name: str
    type: str  # bullish, bearish, neutral
    confidence: float
    index: int
    description: str


def detect_all_patterns(df: pd.DataFrame) -> list[dict]:
    if len(df) < 10:
        return []

    detectors = [
        detect_doji,
        detect_hammer,
        detect_engulfing,
        detect_morning_star,
        detect_evening_star,
        detect_three_white_soldiers,
        detect_three_black_crows,
        detect_harami,
        detect_shooting_star,
        detect_spinning_top,
        detect_marubozu,
        detect_tweezer,
        detect_head_and_shoulders,
        detect_double_top_bottom,
        detect_ascending_triangle,
    ]

    all_patterns = []
    for detector in detectors:
        patterns = detector(df)
        all_patterns.extend(patterns)

    all_patterns.sort(key=lambda p: p["index"], reverse=True)
    return all_patterns


def _body(o, c):
    return abs(c - o)


def _upper_shadow(h, o, c):
    return h - max(o, c)


def _lower_shadow(l, o, c):
    return min(o, c) - l


def _avg_body(df, end, n=10):
    start = max(0, end - n)
    bodies = [_body(df["Open"].iloc[i], df["Close"].iloc[i]) for i in range(start, end)]
    return np.mean(bodies) if bodies else 0.0001


def detect_doji(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(1, len(df)):
        o, h, l, c = df["Open"].iloc[i], df["High"].iloc[i], df["Low"].iloc[i], df["Close"].iloc[i]
        body = _body(o, c)
        total_range = h - l
        if total_range > 0 and body / total_range < 0.1:
            results.append({
                "name": "Doji",
                "type": "neutral",
                "confidence": round(1 - (body / total_range), 2),
                "index": i,
                "description": "Indecision candle - open and close are nearly equal. Watch for trend reversal.",
            })
    return results


def detect_hammer(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(5, len(df)):
        o, h, l, c = df["Open"].iloc[i], df["High"].iloc[i], df["Low"].iloc[i], df["Close"].iloc[i]
        body = _body(o, c)
        lower = _lower_shadow(l, o, c)
        upper = _upper_shadow(h, o, c)
        avg = _avg_body(df, i)

        if body > 0 and lower >= 2 * body and upper <= body * 0.3:
            is_downtrend = df["Close"].iloc[i - 5] > df["Close"].iloc[i - 1]
            if is_downtrend:
                results.append({
                    "name": "Hammer",
                    "type": "bullish",
                    "confidence": round(min(lower / (2 * body), 1.0), 2),
                    "index": i,
                    "description": "Bullish reversal pattern after downtrend. Long lower shadow shows buying pressure.",
                })
    return results


def detect_engulfing(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(1, len(df)):
        o1, c1 = df["Open"].iloc[i - 1], df["Close"].iloc[i - 1]
        o2, c2 = df["Open"].iloc[i], df["Close"].iloc[i]

        if c1 < o1 and c2 > o2 and o2 <= c1 and c2 >= o1:
            results.append({
                "name": "Bullish Engulfing",
                "type": "bullish",
                "confidence": round(min(_body(o2, c2) / max(_body(o1, c1), 0.001), 1.0), 2),
                "index": i,
                "description": "Strong bullish reversal. Current candle completely engulfs previous bearish candle.",
            })
        elif c1 > o1 and c2 < o2 and o2 >= c1 and c2 <= o1:
            results.append({
                "name": "Bearish Engulfing",
                "type": "bearish",
                "confidence": round(min(_body(o2, c2) / max(_body(o1, c1), 0.001), 1.0), 2),
                "index": i,
                "description": "Strong bearish reversal. Current candle completely engulfs previous bullish candle.",
            })
    return results


def detect_morning_star(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(2, len(df)):
        o1, c1 = df["Open"].iloc[i - 2], df["Close"].iloc[i - 2]
        o2, c2 = df["Open"].iloc[i - 1], df["Close"].iloc[i - 1]
        o3, c3 = df["Open"].iloc[i], df["Close"].iloc[i]

        if (c1 < o1 and _body(o2, c2) < _body(o1, c1) * 0.3
                and c3 > o3 and c3 > (o1 + c1) / 2):
            results.append({
                "name": "Morning Star",
                "type": "bullish",
                "confidence": 0.85,
                "index": i,
                "description": "Three-candle bullish reversal. Gap down, small body, then strong bullish close.",
            })
    return results


def detect_evening_star(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(2, len(df)):
        o1, c1 = df["Open"].iloc[i - 2], df["Close"].iloc[i - 2]
        o2, c2 = df["Open"].iloc[i - 1], df["Close"].iloc[i - 1]
        o3, c3 = df["Open"].iloc[i], df["Close"].iloc[i]

        if (c1 > o1 and _body(o2, c2) < _body(o1, c1) * 0.3
                and c3 < o3 and c3 < (o1 + c1) / 2):
            results.append({
                "name": "Evening Star",
                "type": "bearish",
                "confidence": 0.85,
                "index": i,
                "description": "Three-candle bearish reversal. Gap up, small body, then strong bearish close.",
            })
    return results


def detect_three_white_soldiers(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(2, len(df)):
        candles = [(df["Open"].iloc[i - j], df["Close"].iloc[i - j]) for j in range(2, -1, -1)]
        if all(c > o for o, c in candles):
            if candles[1][1] > candles[0][1] and candles[2][1] > candles[1][1]:
                if candles[1][0] > candles[0][0] and candles[2][0] > candles[1][0]:
                    results.append({
                        "name": "Three White Soldiers",
                        "type": "bullish",
                        "confidence": 0.9,
                        "index": i,
                        "description": "Three consecutive bullish candles with higher opens and closes. Strong uptrend signal.",
                    })
    return results


def detect_three_black_crows(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(2, len(df)):
        candles = [(df["Open"].iloc[i - j], df["Close"].iloc[i - j]) for j in range(2, -1, -1)]
        if all(c < o for o, c in candles):
            if candles[1][1] < candles[0][1] and candles[2][1] < candles[1][1]:
                if candles[1][0] < candles[0][0] and candles[2][0] < candles[1][0]:
                    results.append({
                        "name": "Three Black Crows",
                        "type": "bearish",
                        "confidence": 0.9,
                        "index": i,
                        "description": "Three consecutive bearish candles with lower opens and closes. Strong downtrend signal.",
                    })
    return results


def detect_harami(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(1, len(df)):
        o1, c1 = df["Open"].iloc[i - 1], df["Close"].iloc[i - 1]
        o2, c2 = df["Open"].iloc[i], df["Close"].iloc[i]

        if _body(o1, c1) > _body(o2, c2) * 2:
            if min(o2, c2) > min(o1, c1) and max(o2, c2) < max(o1, c1):
                pattern_type = "bullish" if c1 < o1 else "bearish"
                results.append({
                    "name": f"{pattern_type.title()} Harami",
                    "type": pattern_type,
                    "confidence": 0.7,
                    "index": i,
                    "description": f"{pattern_type.title()} reversal - small candle contained within previous large candle.",
                })
    return results


def detect_shooting_star(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(5, len(df)):
        o, h, l, c = df["Open"].iloc[i], df["High"].iloc[i], df["Low"].iloc[i], df["Close"].iloc[i]
        body = _body(o, c)
        upper = _upper_shadow(h, o, c)
        lower = _lower_shadow(l, o, c)

        if body > 0 and upper >= 2 * body and lower <= body * 0.3:
            is_uptrend = df["Close"].iloc[i - 1] > df["Close"].iloc[i - 5]
            if is_uptrend:
                results.append({
                    "name": "Shooting Star",
                    "type": "bearish",
                    "confidence": 0.75,
                    "index": i,
                    "description": "Bearish reversal at top of uptrend. Long upper shadow shows selling pressure.",
                })
    return results


def detect_spinning_top(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(1, len(df)):
        o, h, l, c = df["Open"].iloc[i], df["High"].iloc[i], df["Low"].iloc[i], df["Close"].iloc[i]
        body = _body(o, c)
        upper = _upper_shadow(h, o, c)
        lower = _lower_shadow(l, o, c)
        avg = _avg_body(df, i)

        if body < avg * 0.5 and upper > body and lower > body:
            results.append({
                "name": "Spinning Top",
                "type": "neutral",
                "confidence": 0.6,
                "index": i,
                "description": "Small body with long shadows. Market indecision - potential trend change.",
            })
    return results


def detect_marubozu(df: pd.DataFrame) -> list[dict]:
    results = []
    for i in range(len(df)):
        o, h, l, c = df["Open"].iloc[i], df["High"].iloc[i], df["Low"].iloc[i], df["Close"].iloc[i]
        body = _body(o, c)
        total = h - l
        if total > 0 and body / total > 0.9:
            pattern_type = "bullish" if c > o else "bearish"
            results.append({
                "name": f"{pattern_type.title()} Marubozu",
                "type": pattern_type,
                "confidence": round(body / total, 2),
                "index": i,
                "description": f"Strong {pattern_type} candle with no shadows. High conviction move.",
            })
    return results


def detect_tweezer(df: pd.DataFrame) -> list[dict]:
    results = []
    tolerance = 0.001
    for i in range(1, len(df)):
        h1, l1 = df["High"].iloc[i - 1], df["Low"].iloc[i - 1]
        h2, l2 = df["High"].iloc[i], df["Low"].iloc[i]
        o1, c1 = df["Open"].iloc[i - 1], df["Close"].iloc[i - 1]
        o2, c2 = df["Open"].iloc[i], df["Close"].iloc[i]

        if abs(h1 - h2) / max(h1, 0.01) < tolerance and c1 > o1 and c2 < o2:
            results.append({
                "name": "Tweezer Top",
                "type": "bearish",
                "confidence": 0.75,
                "index": i,
                "description": "Equal highs signal resistance. Bearish reversal pattern.",
            })
        if abs(l1 - l2) / max(l1, 0.01) < tolerance and c1 < o1 and c2 > o2:
            results.append({
                "name": "Tweezer Bottom",
                "type": "bullish",
                "confidence": 0.75,
                "index": i,
                "description": "Equal lows signal support. Bullish reversal pattern.",
            })
    return results


def detect_head_and_shoulders(df: pd.DataFrame) -> list[dict]:
    results = []
    if len(df) < 30:
        return results

    highs = df["High"].values
    for i in range(20, len(df)):
        window = highs[i - 20:i + 1]
        peaks = []
        for j in range(2, len(window) - 2):
            if window[j] > window[j - 1] and window[j] > window[j - 2] and window[j] > window[j + 1] and window[j] > window[j + 2]:
                peaks.append((j, window[j]))

        if len(peaks) >= 3:
            for k in range(len(peaks) - 2):
                left, head, right = peaks[k], peaks[k + 1], peaks[k + 2]
                if head[1] > left[1] and head[1] > right[1]:
                    if abs(left[1] - right[1]) / head[1] < 0.03:
                        results.append({
                            "name": "Head and Shoulders",
                            "type": "bearish",
                            "confidence": 0.8,
                            "index": i,
                            "description": "Classic reversal pattern. Head higher than both shoulders signals trend reversal.",
                        })
                        break
    return results


def detect_double_top_bottom(df: pd.DataFrame) -> list[dict]:
    results = []
    if len(df) < 20:
        return results

    highs = df["High"].values
    lows = df["Low"].values

    for i in range(15, len(df)):
        h_window = highs[i - 15:i + 1]
        l_window = lows[i - 15:i + 1]

        h_peaks = []
        l_troughs = []
        for j in range(2, len(h_window) - 2):
            if h_window[j] > h_window[j - 1] and h_window[j] > h_window[j + 1]:
                h_peaks.append((j, h_window[j]))
            if l_window[j] < l_window[j - 1] and l_window[j] < l_window[j + 1]:
                l_troughs.append((j, l_window[j]))

        if len(h_peaks) >= 2:
            p1, p2 = h_peaks[-2], h_peaks[-1]
            if abs(p1[1] - p2[1]) / max(p1[1], 0.01) < 0.02 and abs(p1[0] - p2[0]) >= 5:
                results.append({
                    "name": "Double Top",
                    "type": "bearish",
                    "confidence": 0.8,
                    "index": i,
                    "description": "Two peaks at similar levels. Resistance confirmed - bearish reversal likely.",
                })

        if len(l_troughs) >= 2:
            t1, t2 = l_troughs[-2], l_troughs[-1]
            if abs(t1[1] - t2[1]) / max(t1[1], 0.01) < 0.02 and abs(t1[0] - t2[0]) >= 5:
                results.append({
                    "name": "Double Bottom",
                    "type": "bullish",
                    "confidence": 0.8,
                    "index": i,
                    "description": "Two troughs at similar levels. Support confirmed - bullish reversal likely.",
                })
    return results


def detect_ascending_triangle(df: pd.DataFrame) -> list[dict]:
    results = []
    if len(df) < 20:
        return results

    for i in range(20, len(df)):
        window_h = df["High"].iloc[i - 15:i + 1].values
        window_l = df["Low"].iloc[i - 15:i + 1].values

        high_std = np.std(window_h[-5:])
        low_slope = np.polyfit(range(len(window_l)), window_l, 1)[0]

        if high_std < np.mean(window_h) * 0.01 and low_slope > 0:
            results.append({
                "name": "Ascending Triangle",
                "type": "bullish",
                "confidence": 0.75,
                "index": i,
                "description": "Flat resistance with rising lows. Bullish breakout pattern.",
            })
    return results

File: backend-rust/src/services/pattern_recognition.rs
//! Candlestick pattern detection — 15 classic patterns.
//!
//! Each detector returns a list of (index, pattern_name, direction, confidence).

use crate::models::Candle;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct PatternResult {
    pub index: usize,
    pub name: String,
    pub direction: PatternDir,
    pub confidence: f64,
    pub description: String,
}

#[derive(Debug, Clone, Copy, Serialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum PatternDir {
    Bullish,
    Bearish,
    Neutral,
}

impl PatternDir {
    pub fn signal(&self) -> f64 {
        match self {
            Self::Bullish => 1.0,
            Self::Bearish => -1.0,
            Self::Neutral => 0.0,
        }
    }
}

/// Detect all patterns on the given candle series.
pub fn detect_all(candles: &[Candle]) -> Vec<PatternResult> {
    let mut results = Vec::new();
    let fns: Vec<fn(&[Candle]) -> Vec<PatternResult>> = vec![
        detect_doji,
        detect_hammer,
        detect_inverted_hammer,
        detect_engulfing,
        detect_morning_star,
        detect_evening_star,
        detect_three_white_soldiers,
        detect_three_black_crows,
        detect_harami,
        detect_piercing_line,
        detect_dark_cloud,
        detect_spinning_top,
        detect_marubozu,
        detect_tweezer,
        detect_hanging_man,
    ];

    for f in fns {
        results.extend(f(candles));
    }
    results.sort_by(|a, b| b.index.cmp(&a.index));
    results
}

fn avg_body(candles: &[Candle]) -> f64 {
    if candles.is_empty() { return 0.001; }
    let sum: f64 = candles.iter().map(|c| c.body()).sum();
    (sum / candles.len() as f64).max(0.001)
}

fn detect_doji(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    if c.len() < 2 { return r; }
    let ab = avg_body(c);
    for i in 0..c.len() {
        if c[i].body() < ab * 0.1 && c[i].range() > ab * 0.5 {
            r.push(PatternResult {
                index: i,
                name: "Doji".into(),
                direction: PatternDir::Neutral,
                confidence: 0.6,
                description: "Indecision — open equals close".into(),
            });
        }
    }
    r
}

fn detect_hammer(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    let ab = avg_body(c);
    for i in 0..c.len() {
        let body = c[i].body();
        let ls = c[i].lower_shadow();
        let us = c[i].upper_shadow();
        if body > 0.0 && ls >= body * 2.0 && us < body * 0.5 && body < ab * 1.5 {
            r.push(PatternResult {
                index: i,
                name: "Hammer".into(),
                direction: PatternDir::Bullish,
                confidence: 0.7,
                description: "Bullish reversal — long lower shadow, small body".into(),
            });
        }
    }
    r
}

fn detect_inverted_hammer(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    let ab = avg_body(c);
    for i in 0..c.len() {
        let body = c[i].body();
        let ls = c[i].lower_shadow();
        let us = c[i].upper_shadow();
        if body > 0.0 && us >= body * 2.0 && ls < body * 0.5 && body < ab * 1.5 {
            r.push(PatternResult {
                index: i,
                name: "Inverted Hammer".into(),
                direction: PatternDir::Bullish,
                confidence: 0.6,
                description: "Potential bullish reversal — long upper shadow".into(),
            });
        }
    }
    r
}

fn detect_engulfing(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    for i in 1..c.len() {
        let prev = &c[i - 1];
        let curr = &c[i];
        if prev.is_bearish()
            && curr.is_bullish()
            && curr.open <= prev.close
            && curr.close >= prev.open
        {
            r.push(PatternResult {
                index: i,
                name: "Bullish Engulfing".into(),
                direction: PatternDir::Bullish,
                confidence: 0.8,
                description: "Strong bullish reversal — current candle engulfs previous".into(),
            });
        }
        if prev.is_bullish()
            && curr.is_bearish()
            && curr.open >= prev.close
            && curr.close <= prev.open
        {
            r.push(PatternResult {
                index: i,
                name: "Bearish Engulfing".into(),
                direction: PatternDir::Bearish,
                confidence: 0.8,
                description: "Strong bearish reversal — current candle engulfs previous".into(),
            });
        }
    }
    r
}

fn detect_morning_star(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    if c.len() < 3 { return r; }
    let ab = avg_body(c);
    for i in 2..c.len() {
        if c[i - 2].is_bearish()
            && c[i - 2].body() > ab
            && c[i - 1].body() < ab * 0.3
            && c[i].is_bullish()
            && c[i].body() > ab
            && c[i].close > c[i - 2].midpoint()
        {
            r.push(PatternResult {
                index: i,
                name: "Morning Star".into(),
                direction: PatternDir::Bullish,
                confidence: 0.85,
                description: "Strong bullish reversal — three-candle pattern".into(),
            });
        }
    }
    r
}

fn detect_evening_star(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    if c.len() < 3 { return r; }
    let ab = avg_body(c);
    for i in 2..c.len() {
        if c[i - 2].is_bullish()
            && c[i - 2].body() > ab
            && c[i - 1].body() < ab * 0.3
            && c[i].is_bearish()
            && c[i].body() > ab
            && c[i].close < c[i - 2].midpoint()
        {
            r.push(PatternResult {
                index: i,
                name: "Evening Star".into(),
                direction: PatternDir::Bearish,
                confidence: 0.85,
                description: "Strong bearish reversal — three-candle pattern".into(),
            });
        }
    }
    r
}

fn detect_three_white_soldiers(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    if c.len() < 3 { return r; }
    for i in 2..c.len() {
        if c[i - 2].is_bullish()
            && c[i - 1].is_bullish()
            && c[i].is_bullish()
            && c[i - 1].close > c[i - 2].close
            && c[i].close > c[i - 1].close
            && c[i - 1].open > c[i - 2].open
            && c[i].open > c[i - 1].open
        {
            r.push(PatternResult {
                index: i,
                name: "Three White Soldiers".into(),
                direction: PatternDir::Bullish,
                confidence: 0.9,
                description: "Very bullish — three consecutive higher closes".into(),
            });
        }
    }
    r
}

fn detect_three_black_crows(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    if c.len() < 3 { return r; }
    for i in 2..c.len() {
        if c[i - 2].is_bearish()
            && c[i - 1].is_bearish()
            && c[i].is_bearish()
            && c[i - 1].close < c[i - 2].close
            && c[i].close < c[i - 1].close
        {
            r.push(PatternResult {
                index: i,
                name: "Three Black Crows".into(),
                direction: PatternDir::Bearish,
                confidence: 0.9,
                description: "Very bearish — three consecutive lower closes".into(),
            });
        }
    }
    r
}

fn detect_harami(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    for i in 1..c.len() {
        let prev = &c[i - 1];
        let curr = &c[i];
        if prev.is_bearish()
            && curr.is_bullish()
            && curr.body() < prev.body()
            && curr.open > prev.close
            && curr.close < prev.open
        {
            r.push(PatternResult {
                index: i,
                name: "Bullish Harami".into(),
                direction: PatternDir::Bullish,
                confidence: 0.65,
                description: "Potential bullish reversal — small body inside previous".into(),
            });
        }
        if prev.is_bullish()
            && curr.is_bearish()
            && curr.body() < prev.body()
            && curr.open < prev.close
            && curr.close > prev.open
        {
            r.push(PatternResult {
                index: i,
                name: "Bearish Harami".into(),
                direction: PatternDir::Bearish,
                confidence: 0.65,
                description: "Potential bearish reversal — small body inside previous".into(),
            });
        }
    }
    r
}

fn detect_piercing_line(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    for i in 1..c.len() {
        if c[i - 1].is_bearish()
            && c[i].is_bullish()
            && c[i].open < c[i - 1].low
            && c[i].close > c[i - 1].midpoint()
            && c[i].close < c[i - 1].open
        {
            r.push(PatternResult {
                index: i,
                name: "Piercing Line".into(),
                direction: PatternDir::Bullish,
                confidence: 0.7,
                description: "Bullish reversal — opens below low, closes above midpoint".into(),
            });
        }
    }
    r
}

fn detect_dark_cloud(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    for i in 1..c.len() {
        if c[i - 1].is_bullish()
            && c[i].is_bearish()
            && c[i].open > c[i - 1].high
            && c[i].close < c[i - 1].midpoint()
            && c[i].close > c[i - 1].open
        {
            r.push(PatternResult {
                index: i,
                name: "Dark Cloud Cover".into(),
                direction: PatternDir::Bearish,
                confidence: 0.7,
                description: "Bearish reversal — opens above high, closes below midpoint".into(),
            });
        }
    }
    r
}

fn detect_spinning_top(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    let ab = avg_body(c);
    for i in 0..c.len() {
        let body = c[i].body();
        let us = c[i].upper_shadow();
        let ls = c[i].lower_shadow();
        if body < ab * 0.3 && us > body && ls > body && c[i].range() > ab * 0.5 {
            r.push(PatternResult {
                index: i,
                name: "Spinning Top".into(),
                direction: PatternDir::Neutral,
                confidence: 0.5,
                description: "Indecision — small body with shadows on both sides".into(),
            });
        }
    }
    r
}

fn detect_marubozu(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    let ab = avg_body(c);
    for i in 0..c.len() {
        let body = c[i].body();
        let us = c[i].upper_shadow();
        let ls = c[i].lower_shadow();
        let shadow_ratio = if body > 0.0 { (us + ls) / body } else { 99.0 };
        if body > ab * 1.5 && shadow_ratio < 0.1 {
            let dir = if c[i].is_bullish() {
                PatternDir::Bullish
            } else {
                PatternDir::Bearish
            };
            r.push(PatternResult {
                index: i,
                name: "Marubozu".into(),
                direction: dir,
                confidence: 0.75,
                description: "Strong conviction — no shadows".into(),
            });
        }
    }
    r
}

fn detect_tweezer(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    if c.len() < 2 { return r; }
    let ab = avg_body(c);
    let tolerance = ab * 0.1;
    for i in 1..c.len() {
        // Tweezer bottom
        if c[i - 1].is_bearish()
            && c[i].is_bullish()
            && (c[i].low - c[i - 1].low).abs() < tolerance
        {
            r.push(PatternResult {
                index: i,
                name: "Tweezer Bottom".into(),
                direction: PatternDir::Bullish,
                confidence: 0.7,
                description: "Bullish reversal — matching lows".into(),
            });
        }
        // Tweezer top
        if c[i - 1].is_bullish()
            && c[i].is_bearish()
            && (c[i].high - c[i - 1].high).abs() < tolerance
        {
            r.push(PatternResult {
                index: i,
                name: "Tweezer Top".into(),
                direction: PatternDir::Bearish,
                confidence: 0.7,
                description: "Bearish reversal — matching highs".into(),
            });
        }
    }
    r
}

fn detect_hanging_man(c: &[Candle]) -> Vec<PatternResult> {
    let mut r = Vec::new();
    let ab = avg_body(c);
    for i in 0..c.len() {
        let body = c[i].body();
        let ls = c[i].lower_shadow();
        let us = c[i].upper_shadow();
        if c[i].is_bearish() && body > 0.0 && ls >= body * 2.0 && us < body * 0.3 && body < ab * 1.5 {
            r.push(PatternResult {
                index: i,
                name: "Hanging Man".into(),
                direction: PatternDir::Bearish,
                confidence: 0.65,
                description: "Bearish reversal — long lower shadow after uptrend".into(),
            });
        }
    }
    r
}
